src/Symfony/Component/ObjectMapper/CachedObjectMapper.php

```diff
@@ -0,0 +1,351 @@
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\ObjectMapper;
+
+use Psr\Container\ContainerInterface;
+use Symfony\Component\Filesystem\Exception\IOException;
+use Symfony\Component\Filesystem\Filesystem;
+use Symfony\Component\ObjectMapper\Exception\MappingException;
+use Symfony\Component\ObjectMapper\Metadata\ObjectMapperMetadataFactoryInterface;
+use Symfony\Component\ObjectMapper\Metadata\ReflectionObjectMapperMetadataFactory;
+use Symfony\Component\PropertyAccess\PropertyAccessorInterface;
+use Symfony\Component\VarExporter\Exception\NotInstantiableTypeException;
+use Symfony\Component\VarExporter\LazyObjectInterface;
+use Symfony\Component\VarExporter\VarExporter;
+
+/**
+ * Cached Object to object mapper that generates and caches compiled mapping functions.
+ *
+ * @author Antoine Bluchet <soyuka@gmail.com>
+ */
+final class CachedObjectMapper implements ObjectMapperInterface, ObjectMapperAwareInterface
+{
+    use ObjectMapperAwareTrait;
+    use ObjectMapperTrait;
+
+    private array $cachedMappings = [];
+    private ?Filesystem $fs = null;
+
+    /**
+     * Tracks recursive references.
+     */
+    private ?\SplObjectStorage $objectMap = null;
+
+    public function __construct(
+        private readonly string $cacheDir,
+        ObjectMapperMetadataFactoryInterface $metadataFactory = new ReflectionObjectMapperMetadataFactory(),
+        ?PropertyAccessorInterface $propertyAccessor = null,
+        ?ContainerInterface $transformCallableLocator = null,
+        ?ContainerInterface $conditionCallableLocator = null,
+        ?ObjectMapperInterface $objectMapper = null,
+        private readonly bool $debug = false,
+    ) {
+        $this->metadataFactory = $metadataFactory;
+        $this->propertyAccessor = $propertyAccessor;
+        $this->transformCallableLocator = $transformCallableLocator;
+        $this->conditionCallableLocator = $conditionCallableLocator;
+        $this->objectMapper = $objectMapper;
+    }
+
+    public function map(object $source, object|string|null $target = null): object
+    {
+        $objectMapInitialized = false;
+        if (null === $this->objectMap) {
+            $this->objectMap = new \SplObjectStorage();
+            $objectMapInitialized = true;
+        }
+
+        $sourceClass = $source::class;
+        $targetClass = $this->resolveTargetClass($source, $target);
+
+        if (!$targetClass) {
+            throw new MappingException(\sprintf('Mapping target not found for source "%s".', get_debug_type($source)));
+        }
+
+        if (\is_string($target) && !class_exists($target)) {
+            throw new MappingException(\sprintf('Mapping target class "%s" does not exist for source "%s".', $target, get_debug_type($source)));
+        }
+
+        $cacheKey = str_replace(['\\', '/'], '_', $sourceClass.'-to-'.$targetClass);
```

Comment:
this might have to be tweaked on windows where the full path cannot be longer than 258 chars
on Linux, file names are limited to 255 chars
here is a proposal:
```suggestion
        $cacheKey = preg_replace_callback('^((?:[^\\\\]++\\\\)*+)([^\\\\]++)-to-((?:[^\\\\]++\\\\)*+)([^\\\\]++)$', static fn ($m) => $m[2].'-to-'.$m[4].'-'.hash('xxh128', $m[1].'-'.$m[3]),$sourceClass.'-to-'.$targetClass);
```